### DISCUSSION TITLE
ERA5 forcing python codes

### DISCUSSION THREAD

**Discussion by ehsanj25** [NCAR/hrldas#111] [Q&A] [answered]
After following the steps provided in [this tutorial](https://github.com/CharlesZheZhang/NoahMP_Tutorial/blob/main/Note5_2D_ERA5-Land_forcing.ipynb), I have succeeded in generating the majority of the input forcings. Nevertheless, I have encountered some minor issues that I wish to discuss:

1- The snow depth (SD) output, derived from the initialization code, perpetually returns empty.

2- I still do not understand why there is a need to decumulate the data, given that it reflects the accumulation of the parameter within the one-hour period immediately preceding the validity date and time.

3- Within the calculate_Q_era5.py script, the cdo line fails to operate with my output file. I consistently encounter error messages stating, Variable >var168< not found!. When I attempt to replace it with the variable name I obtained from cdo showname ERA5_2D_21010100, which is 2d, I encounter another error: cdo expr (Abort): No output variable found! Has anyone else experienced this issue?

I would appreciate any insights or suggestions to address these issues.

**Comment by tslin2** [comment, marked as answer]
maybe you can give try to another processing code as follows and make sure it is ERA5-Land data.
https://github.com/NCAR/hrldas/blob/f550db9f22c11acbf7aec607904bc22781425e15/hrldas/HRLDAS_forcing/run/examples/ERA5/Prepare_ERA5-Land_Forcing.sh

For solar and thermal radiation downwards and total precipitation, ERA5-Land is accumulated from the beginning of the forecast time to the end of the forecast step, so decumulate is needed.
https://cds.climate.copernicus.eu/cdsapp#!/dataset/reanalysis-era5-land?tab=overview


If you use ERA5, then there is no need to do that. ERA5 and ERA5-Land is different.

**Reply by tslin2** [reply]
2m dewpoint temperature is required to calculate calculate_Q_era5.py, which seems it was missing in your file.

**Reply by ehsanj25** [reply]
Hi Tzu-Shun,

Thank you for your guidance.

[This](https://github.com/NCAR/hrldas/blob/f550db9f22c11acbf7aec607904bc22781425e15/hrldas/HRLDAS_forcing/run/examples/ERA5/Prepare_ERA5-Land_Forcing.sh) code worked for me, but I'd like to highlight a few prerequisites that were essential for me:

- **Installation of ecCodes and cdo**: It was necessary to have both **ecCodes** and **cdo (with ecCodes support)** installed on my system.
- **System Path Addition**: Additionally, ensuring that these were added to the system path is important

Most of the functions, such as **grib_copy** or **codes_split_file**, utilized in the bash script and Python codes are derived from these libraries, which might be a useful note for future reference.

Moreover, a clarification regarding the SD parameter: It appears that the model is seeking the Snow Water Equivalent (SWE) rather than Snow Depth (SD). Once I utilized the SWE parameter, the issue with the initialization code resolved.

**Comment by CharlesZheZhang** [comment]
Hello Ehsan,

To your second question on the de-accumulation, you can find more information on the [ERA5-Land online documentation](https://confluence.ecmwf.int/display/CKB/ERA5-Land%3A+data+documentation), see Accumulations: 

''Please, note that the convention for accumulations used in ERA5-Land differs with that for ERA5. The accumulations in the short forecasts of ERA5-Land (with hourly steps from 01 to 24) are treated the same as those in ERA-Interim or ERA-Interim/Land, i.e., they are accumulated from the beginning of the forecast to the end of the forecast step. For example, runoff at day=D, step=12 will provide runoff accumulated from day=D, time=0 to day=D, time=12. The maximum accumulation is over 24 hours, i.e., from day=D, time=0 to day=D+1,time=0 (step=24).''
The accumulations starts from the first forecast timestep at 00 hour to the end of the forecast step, that's why we need to deaccumulate three variables (precipitation, short and longwave radiation).

Also thanks for the note for SD and SWE, will correct it in the extract initial condition code.

Many thanks,
Zhe

**Reply by CharlesZheZhang** [reply]
Hello Ehsan, I want to follow up with the SD issue you had with the ERA5-Land initial condition. 
I check with the ERA5-Land data documentation, the naming for snow depth and snow depth water equivalent is quite confusing:
[https://confluence.ecmwf.int/display/CKB/ERA5-Land%3A+data+documentation](url)

The variable **snow depth** is with the short name **sde** 
while the variable **snow depth water equivalent** is with short name **sd**.

So, after the initial data extracted from the ERA5-Land file, using the python code, it will be named after its short name **ERA5_SD_<date>**, but it represents **snow depth water equivalent**.

Therefore, use this filename **ERA5_SD_<date>** in create_forcing.exe namelist **WEASDfile_template** should be OK. 

Thank you once again for reminding me the SD variable name, I will update this in the ERA5-Land documentation.

**Comment by ehsanj25** [comment]
Hi Zhe,

Of course, thanks you for following up on this. Yes, the short names are
kind of misleading and it is great that you are highlighting this in the
documentation.

Best,
Ehsan



On Tue, 24 Oct 2023 at 18:20 Zhe Zhang ***@***.***> wrote:

> Hello Ehsan, I want to follow up with the SD issue you had with the
> ERA5-Land initial condition.
> I check with the ERA5-Land data documentation, the naming for snow depth
> and snow depth water equivalent is quite confusing:
> https://confluence.ecmwf.int/display/CKB/ERA5-Land%3A+data+documentation
> <http://url>
>
> The variable *snow depth* is with the short name *sde*
> while the variable *snow depth water equivalent* is with short name *sd*.
>
> So, after the initial data extracted from the ERA5-Land file, using the
> python code, it will be named after its short name *ERA5_SD_*, but it
> represents *snow depth water equivalent*.
>
> Therefore, use this filename *ERA5_SD_* in create_forcing.exe namelist
> *WEASDfile_template* should be OK.
>
> Thank you once again for reminding me the SD variable name, I will update
> this in the ERA5-Land documentation.
>
> —
> Reply to this email directly, view it on GitHub
> <https://github.com/NCAR/hrldas/discussions/111#discussioncomment-7375015>,
> or unsubscribe
> <https://github.com/notifications/unsubscribe-auth/ADSI4CYUUHSYMXH67J2I2NDYBA5KXAVCNFSM6AAAAAA5J67Z4OVHI2DSMVQWIX3LMV43SRDJONRXK43TNFXW4Q3PNVWWK3TUHM3TGNZVGAYTK>
> .
> You are receiving this because you authored the thread.Message ID:
> ***@***.***>
>
-- 

*Ehsan** Jalilvand*

Postdoctoral Research Associate

Biosystem and Agricultural Engineering Department

Michigan State University

*Personal website: https://www.ehsanjalilvand.ir
<https://www.ehsanjalilvand.ir/>*

<https://scholar.google.com/citations?user=RsLEzG0AAAAJ&hl=en>
<https://twitter.com/ehsan1325>
<https://www.youtube.com/channel/UC6SyjY2pXoI0UK1tfY7uxcw>
<https://www.researchgate.net/profile/Ehsan-Jalilvand>
<https://www.linkedin.com/in/ehsan-jalilvand-1699534a/>
